components/tidb_query/src/codec/row/v2/decoder.rs

```diff
@@ -0,0 +1,332 @@
+use crate::codec::{
+    mysql::{self, DecimalDecoder, JsonDecoder, Time},
+    Datum, Result,
+};
+use crate::expr::EvalContext;
+use codec::number::NumberCodec;
+use codec::prelude::*;
+use tidb_query_datatype::FieldTypeAccessor;
+use tikv_util::collections::HashMap;
+use tipb::ColumnInfo;
+
+pub trait RowDecoder: NumberDecoder + IntDecoder {
+    // | version | flag | number_of_not_null_values | number_of_null_values | not_null_value_ids | null_value_ids | value_offsets | values
+    fn decode_row(
+        &mut self,
+        ctx: &EvalContext,
+        cols: &HashMap<i64, ColumnInfo>,
+    ) -> Result<HashMap<i64, Datum>> {
+        let mut row = HashMap::with_capacity_and_hasher(cols.len(), Default::default());
+        if self.read_u8()? != super::CODEC_VERSION {
+            panic!("version not matched");
+        }
+        let is_large_id = self.read_u8()? == 1;
+
+        // read ids count
+        let not_null_cnt = self.read_u16_le()? as usize;
+        let null_cnt = self.read_u16_le()? as usize;
+
+        let ids: Vec<i64>;
+        let offsets: Vec<usize>;
+        if is_large_id {
+            ids = self.read_ints_le(not_null_cnt + null_cnt, u32::to_i64)?;
+            offsets = self.read_ints_le(not_null_cnt, u32::to_usize)?;
+        } else {
+            ids = self.read_ints_le(not_null_cnt + null_cnt, u8::to_i64)?;
+            offsets = self.read_ints_le(not_null_cnt, u16::to_usize)?;
+        };
+
+        let values = self.bytes();
+        let (not_null_ids, null_ids) = ids.split_at(not_null_cnt);
+        for (id, tp) in cols {
+            if let Ok(idx) = not_null_ids.binary_search(id) {
+                let offset = offsets
+                    .get(idx)
+                    .expect("should always have value in offsets");
```

Comment:
It's better to returns error to the caller to avoid TiKV crash due to the wrong push down parameters.